### ISSUE TITLE
Superuser can't add model

### ISSUE THREAD

**Issue by sebastienpattyn93** [juju/python-libjuju#122]
Hi I'm running libjuju from source and I created a new User in my environment called sebastien and granted him Superuser access on my controller. I changed his password to 'testpass'. this is my output when I execute `juju users`:

```
Controller: testcontroller

Name       Display name  Access     Date created    Last connection
admin*     admin         superuser  2017-05-09      10 seconds ago
sebastien                superuser  18 minutes ago  11 minutes ago
```
Now I try to establish a connection with libjuju with this user and add a model to my controller by running the following code:

```
async def run(): 
    try:
        controller = Controller()
        await controller.connect(
            url,
            'sebastien',
            'testpass')
        controller.add_model('testmodel')


    except JujuAPIError as e:
        print('error: {}'.format(e.message))
```
but I get the following error:
```
App 18417 stderr:   File "/usr/local/lib/python3.5/dist-packages/juju/controller.py", line 100, in add_model
App 18417 stderr:     region
App 18417 stderr:   File "/usr/local/lib/python3.5/dist-packages/juju/client/facade.py", line 392, in wrapper
App 18417 stderr:     reply = await f(*args, **kwargs)
App 18417 stderr:   File "/usr/local/lib/python3.5/dist-packages/juju/client/_client2.py", line 3358, in CreateModel
App 18417 stderr:     reply = await self.rpc(msg)
App 18417 stderr:   File "/usr/local/lib/python3.5/dist-packages/juju/client/facade.py", line 514, in rpc
App 18417 stderr:     result = await self.connection.rpc(msg, encoder=TypeEncoder)
App 18417 stderr:   File "/usr/local/lib/python3.5/dist-packages/juju/client/connection.py", line 244, in rpc
App 18417 stderr:     raise JujuAPIError(result)
App 18417 stderr: juju.errors.JujuAPIError: no credential specified

```
my credentials look like this:
```
Cloud   Credentials
google  googlecontroller
```

When I add a credential_name and cloud_name by changing my code to this
```
    controller.add_model('testmodel', cloud_name='google', credential_name='googlecontroller')
```
I get the following error:
```
App 18711 stderr:   File "/usr/local/lib/python3.5/dist-packages/juju/controller.py", line 100, in add_model
App 18711 stderr:     region
App 18711 stderr:   File "/usr/local/lib/python3.5/dist-packages/juju/client/facade.py", line 392, in wrapper
App 18711 stderr:     reply = await f(*args, **kwargs)
App 18711 stderr:   File "/usr/local/lib/python3.5/dist-packages/juju/client/_client2.py", line 3358, in CreateModel
App 18711 stderr:     reply = await self.rpc(msg)
App 18711 stderr:   File "/usr/local/lib/python3.5/dist-packages/juju/client/facade.py", line 514, in rpc
App 18711 stderr:     result = await self.connection.rpc(msg, encoder=TypeEncoder)
App 18711 stderr:   File "/usr/local/lib/python3.5/dist-packages/juju/client/connection.py", line 244, in rpc
App 18711 stderr:     raise JujuAPIError(result)
App 18711 stderr: juju.errors.JujuAPIError: getting credential: cloud credential "google/sebastien/googlecontroller" not found
```
EDIT: 
I tried adding `owner= tag.user('admin')` or `owner=tag.user('sebastien')` when I add_model but then I got errors that the I had no permission(admin) or the credentials could not be found(sebastien). Do I have to login first with my 'Sebastien' user first and add the same credentials with a CLI call?

**Comment by sebastienpattyn93** [comment]
I added a model manually 'juju add-model test' and saw the following:
`Uploading credential 'google/sebastien/googlecontroller' to controller`

Is there a way to do this in libjuju?

**Comment by johnsca** [comment]
Apologies for the delayed response.  As you found, the credentials have to be added to the controller per-user.  Unfortunately, this isn't implemented yet in the libjuju object layer.  I believe the relevant low-level API call would be `CloudFacade.UpdateCredentials`.  Note that once the model is added, you will also need a discharged macaroon to connect to it, and libjuju doesn't currently support discharging macaroons, either, though I believe the Python bakery library is in progress.

**Comment by sebastienpattyn93** [comment]
Thank you for your response. I'll check the `CloudFacade.UpdateCredentials` to find a solution.

**Comment by johnsca** [comment]
@sebastienpattyn93 I'm happy to help try to get this functionality added to libjuju.  Feel free to ping me on IRC (cory_fu) if you have any questions.

For reference, it looks like the bit in the Juju CLI code that needs to be replicated is: https://github.com/juju/juju/blob/develop/cmd/juju/controller/addmodel.go#L224-L229

**Comment by johnsca** [comment]
As I mentioned previously, you're also going to run into issues connecting to the model once added due to needing to discharge a macaroon for the new model.  The [py-macaroon-bakery](https://github.com/go-macaroon-bakery/py-macaroon-bakery) should be what we need to add support for that, with [PR #2](https://github.com/go-macaroon-bakery/py-macaroon-bakery/pull/2) there possibly being sufficient to include that functionality in libjuju.

**Comment by sebastienpattyn93** [comment]
I might start writing the solution next week probably. I'll keep you up to date with the changes or issues I encounter. Thanks for the extra documentation

**Comment by johnsca** [comment]
I think this should be fixed with #168 but can you please confirm?

**Comment by mitechie** [comment]
The good news is that the new discharge ability should unblock this so that the updatecredentials calls can be used to prepare the creds for adding the model?

**Comment by sebastienpattyn93** [comment]
The workaround I found and that is working for version 0.6.1:

```
    cloud_facade = client.CloudFacade.from_connection(controller.connection)
    cloud_cred = client.UpdateCloudCredential(
        client.CloudCredential(credential, 'google'),
        tag.credential('google', username, credential_name)
    )
    await cloud_facade.UpdateCredentials([cloud_cred])
```

Where the credential is the credential in the right format to be added to the cloud, and controller is a Controller instance that is connected.

I have not had the time yet to check the new functionality in the latest libjuju version, but I'll be able to check a lot of the new functionalities next week. I'll keep you up to date in this post. Thank you for looking in to this.

**Comment by sebastienpattyn93** [comment]
I got a new issue on this. When I try to add a model with the gui I got the following error:

`cannot create model: failed to open environ: invalid character '/' looking for beginning of value` 

while I didn't got the error after I used a CLI call to add a model or when I used a credential that was added to the GUI.

**Comment by juanmanuel-tirado** [comment]
This issue has expired.
